dali/pipeline/operators/decoder/cache/cached_decoder_impl.cc

```diff
@@ -29,23 +34,45 @@ CachedDecoderImpl::CachedDecoderImpl(const OpSpec& spec)
     const bool cache_debug = spec.GetArgument<bool>("cache_debug");
     cache_ = ImageCacheFactory::Instance().Get(
       device_id_, cache_type, cache_size, cache_debug, cache_threshold);
+
+    use_batch_copy_kernel_ = spec.GetArgument<bool>("cache_batch_copy");
+    auto batch_size = spec.GetArgument<int>("batch_size");
+    const size_t kMaxSizePerBlock = 1<<18;  // 256 kB per block
+    scatter_gather_.reset(new kernels::ScatterGatherGPU(
+      kMaxSizePerBlock, cache_size, batch_size));
   }
 }
 
 bool CachedDecoderImpl::CacheLoad(const std::string& file_name,
-                                  uint8_t* output_data,
+                                  uint8_t *output_data,
                                   cudaStream_t stream) {
   if (!cache_ || file_name.empty())
     return false;
   return cache_->Read(file_name, output_data, stream);
 }
 
+
+bool CachedDecoderImpl::DeferCacheLoad(const std::string& file_name, uint8_t *output_data) {
```

Comment:
```suggestion
bool CachedDecoderImpl::DeferCacheLoad(const std::string& file_name, uint8_t *output_data, cudaStream_t stream) {
```
and propagate it to the scatter gather kernel